client/session.go

```diff
@@ -1224,79 +1246,88 @@ func (s *session) FetchBlocksMetadataFromPeers(
 	var (
 		metadataCh = make(chan blocksMetadata, blocksMetadataChannelInitialCapacity)
 		errCh      = make(chan error, 1)
-		m          = s.streamFromPeersMetricsForShard(shard, resultTypeMetadata)
+		m          = s.newPeerMetadataStreamingProgressMetrics(shard, resultTypeMetadata)
 	)
 
 	go func() {
-		errCh <- s.streamBlocksMetadataFromPeers(namespace, shard, peers,
-			start, end, metadataCh, m)
+		errCh <- s.streamBlocksMetadataFromPeers(
+			namespace, shard, peers, start, end, metadataCh, m, FetchBlocksMetadataEndpointV1)
 		close(metadataCh)
 		close(errCh)
 	}()
 
 	return newMetadataIter(metadataCh, errCh), nil
 }
 
+// FetchBootstrapBlocksFromPeers will fetch the specified blocks from peers for
+// bootstrapping purposes. Refer to peer_bootstrapping.md for more details.
 func (s *session) FetchBootstrapBlocksFromPeers(
 	nsMetadata namespace.Metadata,
 	shard uint32,
 	start, end time.Time,
 	opts result.Options,
+	version FetchBlocksMetadataEndpointVersion,
 ) (result.ShardResult, error) {
+	if version != FetchBlocksMetadataEndpointV1 && version != FetchBlocksMetadataEndpointV2 {
```

Comment:
Maybe we should move this closer to the definition of:

Reply:
In fact, maybe we don't need to do this here.

Below you can do

Reply:
I think waiting until we reach the streamFn portion of the code is too late because the streamBlocksMetadataFromPeers goroutine is not pre-emptible and will just run until its done.

How about I replace the lines:

```golang
if version != FetchBlocksMetadataEndpointV1 && version != FetchBlocksMetadataEndpointV2 {
		return nil, errInvalidFetchBlocksMetadataVersion
	}
```

with the new IsValidFetchBlocksMetadataEndpoint to remove the code duplication? I can also add the switch statement as specified as an extra sanity check